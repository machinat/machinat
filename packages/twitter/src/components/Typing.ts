/* eslint-disable import/prefer-default-export */
import { makeUnitSegment, UnitSegment } from '@sociably/core/renderer';
import TwitterChat from '../Chat';
import makeTwitterComponent from '../utils/makeTwitterComponent';
import { TwitterSegmentValue, TwitterComponent } from '../types';

/**
 * Display typing action in the direct message chat
 * @category Component
 * @props {@link TypingProps}
 * @guides Check official [reference](https://developer.twitter.com/en/docs/twitter-api/v1/direct-messages/typing-indicator-and-read-receipts/api-reference/new-typing-indicator).
 */
export const Typing: TwitterComponent<
  {},
  UnitSegment<TwitterSegmentValue>
> = makeTwitterComponent(function Typing(node, path) {
  return [
    makeUnitSegment(node, path, {
      type: 'dm',
      request: {
        method: 'POST',
        href: '1.1/direct_messages/indicate_typing.json',
        params: { recipient_id: '' },
      },
      accomplishRequest: (chat: TwitterChat, request) => {
        // eslint-disable-next-line no-param-reassign
        request.params.recipient_id = chat.userId;
        return request;
      },
      mediaSources: null,
    }),
  ];
});
