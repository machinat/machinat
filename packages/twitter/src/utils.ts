import { annotateNativeComponent } from '@machinat/core/renderer';
import TwitterUser from './User';
import TwitterChat from './Chat';
import TweetTarget from './TweetTarget';
import { TWITTER } from './constant';
import type {
  RawEntities,
  UrlEntity,
  MentionEntity,
  RawMedia,
  Media,
  MediaSource,
  DirectMessageSegmentValue,
  TweetSegmentValue,
} from './types';

export const makeTwitterComponent = annotateNativeComponent(TWITTER);

export const polishMentionEntity = (
  raw: RawEntities['user_mentions'][number]
): MentionEntity => ({
  user: new TwitterUser(raw.id_str),
  id: raw.id_str,
  name: raw.name,
  screenName: raw.screen_name,
  indices: raw.indices,
});

export const polishUrlEntity = (
  raw: RawEntities['urls'][number]
): UrlEntity => ({
  url: raw.url,
  displayUrl: raw.display_url,
  expandedUrl: raw.expanded_url,
  indices: raw.indices,
});

export const polishMedia = (raw: RawMedia): Media => {
  const videoInfo = raw.video_info;
  return {
    id: raw.id_str,
    type: raw.type,
    inlineUrl: raw.url,
    displayUrl: raw.display_url,
    expandedUrl: raw.expanded_url,
    indices: raw.indices,
    downloadUrl: raw.media_url_https,
    sizes: raw.sizes,
    sourceStatusId: raw.source_status_id_str,
    videoInfo: videoInfo
      ? {
          aspectRatio: videoInfo.aspect_ratio,
          durationMs: videoInfo.duration_millis,
          variants: videoInfo.variants.map((rawVariant) => ({
            url: rawVariant.url,
            bitrate: rawVariant.bitrate,
            contentType: rawVariant.content_type,
          })),
        }
      : null,
    additionalMediaInfo: raw.additional_media_info || null,
  } as Media;
};

export const annotateDmMedia = (media: MediaSource): MediaSource =>
  media.sourceType === 'id'
    ? media
    : {
        ...media,
        parameters: {
          ...media.parameters,
          media_category:
            media.parameters.media_category ||
            (media.type === 'photo'
              ? 'dm_image'
              : media.type === 'video'
              ? 'dm_video'
              : media.type === 'animated_gif'
              ? 'dm_gif'
              : undefined),
        },
      };

export const annotateTweetMedia = (media: MediaSource): MediaSource =>
  media.sourceType === 'id'
    ? media
    : {
        ...media,
        parameters: {
          ...media.parameters,
          media_category:
            media.parameters.media_category ||
            (media.type === 'photo'
              ? 'tweet_image'
              : media.type === 'video'
              ? 'tweet_video'
              : media.type === 'animated_gif'
              ? 'tweet_gif'
              : undefined),
        },
      };

export const createDmSegmentValue = (
  text?: string,
  media?: MediaSource
): DirectMessageSegmentValue => ({
  type: 'dm',
  request: {
    method: 'POST',
    href: '1.1/direct_messages/events/new.json',
    parameters: {
      event: {
        type: 'message_create',
        message_create: {
          target: { recipient_id: '' },
          message_data: {
            text,
            attachment: media
              ? {
                  type: 'media',
                  media: { id: '' },
                }
              : undefined,
          },
        },
      },
    },
  },
  accomplishRequest: (target: TwitterChat, request, mediaResults) => {
    // eslint-disable-next-line no-param-reassign
    request.parameters.event.message_create.target.recipient_id = target.id;

    if (mediaResults) {
      // eslint-disable-next-line no-param-reassign, prefer-destructuring
      request.parameters.event.message_create.message_data.attachment.media.id =
        mediaResults[0];
    }
    return request;
  },
  mediaSources: media ? [annotateDmMedia(media)] : null,
});

export const createTweetSegmentValue = (
  parameters: Record<string, any>,
  mediaSources?: null | MediaSource[],
  dmLink?:
    | boolean
    | string
    | { recipientId?: string; text?: string; welcomeMessageId?: string }
): TweetSegmentValue => ({
  type: 'tweet',
  request: {
    method: 'POST',
    href: '2/tweets',
    parameters,
  },
  mediaSources: mediaSources ? mediaSources.map(annotateTweetMedia) : null,
  accomplishRequest: (target: TweetTarget, request, mediaIds) => {
    const params = request.parameters;
    return {
      ...request,
      parameters: {
        ...request.parameters,
        reply: target.tweetId
          ? {
              in_reply_to_tweet_id: target.tweetId,
              exclude_reply_user_ids: params.reply?.exclude_reply_user_ids,
            }
          : undefined,
        media: mediaIds
          ? {
              media_ids: mediaIds,
              tagged_user_ids: params.media?.tagged_user_ids,
            }
          : undefined,
        direct_message_deep_link:
          typeof dmLink === 'string'
            ? dmLink
            : typeof dmLink === 'object'
            ? `https://twitter.com/messages/compose?recipient_id=${
                dmLink.recipientId || target.agentId
              }${
                dmLink.welcomeMessageId
                  ? `&welcome_message_id=${dmLink.welcomeMessageId}`
                  : ''
              }${dmLink.text ? `&text=${encodeURIComponent(dmLink.text)}` : ''}`
            : dmLink
            ? `https://twitter.com/messages/compose?recipient_id=${target.agentId}`
            : undefined,
      },
    };
  },
});

let lastTime = 0;
let count = 0;
export const getTimeId = (): string => {
  const now = Date.now();
  if (lastTime === now) {
    count += 1;
  } else {
    lastTime = now;
    count = 0;
  }
  return `${now}-${count}`;
};
