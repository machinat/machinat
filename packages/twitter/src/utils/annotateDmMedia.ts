import type { MediaSource, MediaType } from '../types.js';

const annotateDmMedia = (
  mediaType: MediaType,
  source: MediaSource,
): MediaSource =>
  source.type === 'id'
    ? source
    : {
        ...source,
        params: {
          ...source.params,
          media_category:
            source.params.media_category ||
            (mediaType === 'photo'
              ? 'dm_image'
              : mediaType === 'video'
              ? 'dm_video'
              : mediaType === 'animated_gif'
              ? 'dm_gif'
              : undefined),
        },
      };

export default annotateDmMedia;
