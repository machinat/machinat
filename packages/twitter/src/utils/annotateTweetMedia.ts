import type { MediaSource, MediaType } from '../types.js';

const annotateTweetMedia = (
  mediaType: MediaType,
  source: MediaSource,
): MediaSource =>
  source.type === 'id'
    ? source
    : {
        ...source,
        params: {
          ...source.params,
          media_category:
            source.params.media_category ||
            (mediaType === 'photo'
              ? 'tweet_image'
              : mediaType === 'video'
              ? 'tweet_video'
              : mediaType === 'animated_gif'
              ? 'tweet_gif'
              : undefined),
        },
      };

export default annotateTweetMedia;
