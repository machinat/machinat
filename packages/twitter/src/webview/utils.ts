import type { ContextDetails } from '@sociably/auth';
import TwitterChat from '../Chat';
import TwitterUser from '../User';
import UserProfile from '../UserProfile';
import type { TwitterAuthContext, TwitterAuthData } from './types';

// eslint-disable-next-line import/prefer-default-export
export const getAuthContextDetails = ({
  agent: agentId,
  user: { id: userId, data: rawUserData },
}: TwitterAuthData): ContextDetails<TwitterAuthContext> => {
  return {
    agentId,
    channel: new TwitterUser(agentId),
    user: new TwitterUser(userId, rawUserData),
    thread: new TwitterChat(agentId, userId),
    userProfile: new UserProfile(rawUserData),
  };
};
