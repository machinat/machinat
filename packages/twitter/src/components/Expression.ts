import type { MachinatNode } from '@machinat/core';
import { formatNode } from '@machinat/core/utils';
import { OutputSegment } from '@machinat/core/renderer';
import { makeTwitterComponent, createDmSegmentValue } from '../utils';
import {
  TwitterSegmentValue,
  TwitterComponent,
  TwitterApiRequest,
} from '../types';

/**
 * @category Props
 */
export type ExpressionProps = {
  /** Direct messages content  */
  children: MachinatNode;
  /** Quick replies to be attached after the messages. Should contain only {@link QuickReply} */
  quickReplies?: MachinatNode;
  /** The custome profile to send the messages with */
  customProfileId?: string;
};

/**
 * Send direct messages with metadata
 * @category Component
 * @props {@link ExpressionProps}
 * @guides Check official [guides](https://developer.twitter.com/en/docs/twitter-api/v1/direct-messages/api-features).
 */
export const Expression: TwitterComponent<
  ExpressionProps,
  OutputSegment<TwitterSegmentValue>
> = makeTwitterComponent(async function Expression(node, path, render) {
  const { children, quickReplies, customProfileId } = node.props;

  const [messageSegments, quickRepliesSegments] = await Promise.all([
    render<TwitterSegmentValue>(children, '.children'),
    render(quickReplies, '.quickReplies'),
  ]);

  const outputSegment: OutputSegment<TwitterSegmentValue>[] = [];
  let lastCreateDmRequest: undefined | TwitterApiRequest;

  for (const segment of messageSegments || []) {
    let createDmRequest: undefined | TwitterApiRequest;

    if (segment.type === 'part') {
      throw new TypeError(
        `${formatNode(segment.node)} can't be sent directly in <Expression/>`
      );
    }
    if (segment.type === 'text') {
      const dmSegValue = createDmSegmentValue(segment.value);
      createDmRequest = dmSegValue.request;

      outputSegment.push({
        type: 'unit',
        node: segment.node,
        path: segment.path,
        value: dmSegValue,
      });
    } else if (segment.type === 'unit' || segment.type === 'raw') {
      const { value: segValue } = segment;

      if (segValue.type === 'tweet') {
        throw new TypeError(
          `${formatNode(segment.node)} can't be sent in <Expression/>`
        );
      }
      if (
        segValue.type === 'dm' &&
        segValue.request.href === '1.1/direct_messages/events/new.json'
      ) {
        createDmRequest = segValue.request;
        outputSegment.push(segment);
      } else if (segValue.type === 'media') {
        const dmSegValue = createDmSegmentValue(undefined, segValue.media);
        createDmRequest = dmSegValue.request;

        outputSegment.push({
          type: 'unit',
          node: segment.node,
          path: segment.path,
          value: dmSegValue,
        });
      } else {
        outputSegment.push(segment);
      }
    } else if (segment.type !== 'break') {
      outputSegment.push(segment);
    }

    if (createDmRequest) {
      const dmOptions = createDmRequest.parameters.event.message_create;

      if (!dmOptions.custom_profile_id) {
        dmOptions.custom_profile_id = customProfileId;
      }
      lastCreateDmRequest = createDmRequest;
    }
  }

  if (quickRepliesSegments) {
    if (!lastCreateDmRequest) {
      throw new TypeError(
        'no message content available to attach quick replies'
      );
    }
    lastCreateDmRequest.parameters.event.message_create.message_data.quick_reply =
      {
        type: 'options',
        options: quickRepliesSegments.map(({ value }) => value),
      };
  }

  return outputSegment.length > 0 ? outputSegment : null;
});
