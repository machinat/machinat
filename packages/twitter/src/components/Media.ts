import { NativeElement } from '@machinat/core';
import { makeUnitSegment, UnitSegment } from '@machinat/core/renderer';
import { makeTwitterComponent } from '../utils';
import { TwitterSegmentValue, TwitterComponent, MediaType } from '../types';

/**
 * @category Props
 */
export type MediaProps = {
  /** Set to true if media asset will be reused for multiple Direct Messages. Default is false */
  shared?: boolean;
  /** The tag to store media asset with */
  assetTag?: string;
  /** A string enum value which identifies a media usecase. This identifier is used to enforce usecase specific constraints (e.g. file size, video duration) and enable advanced features */
  mediaCategory?: string;
  /** A comma-separated list of user IDs to set as additional owners allowed to use the returned media_id in Tweets or Cards. Up to 100 additional owners may be specified */
  additionalOwners?: string[];
  /** The media file url */
  url?: string;
  /** The uploaded media id */
  mediaId?: string;
  /** The file content to be upload */
  file?: Buffer | NodeJS.ReadableStream;
  /** The content type of the file. Required when uploading using `file` */
  fileType?: string;
  /** The file size in bytes. Required when uploading using `file` */
  fileSize?: number;
};

const makeMediaComponent = (name: string, type: MediaType) =>
  makeTwitterComponent(
    {
      [name](
        node: NativeElement<MediaProps, TwitterComponent<MediaProps>>,
        path: string
      ): UnitSegment<TwitterSegmentValue>[] {
        const {
          shared,
          assetTag,
          url,
          mediaId,
          file,
          fileSize,
          fileType,
          mediaCategory,
          additionalOwners,
        } = node.props;

        if (
          !(url || mediaId || file) ||
          (url && mediaId) ||
          (url && file) ||
          (file && mediaId)
        ) {
          throw new TypeError(
            'there should be exactly one of "url", "mediaId" or "file" prop'
          );
        }
        if (file && (!fileType || !fileSize)) {
          throw new TypeError(
            '"fileType" and "fileSize" props are required when using "file" source'
          );
        }

        const parameters = {
          shared: shared ? 'true' : undefined,
          media_category: mediaCategory,
          additional_owners: additionalOwners
            ? additionalOwners.join(',')
            : undefined,
        };

        return [
          makeUnitSegment(node, path, {
            type: 'media',
            media: mediaId
              ? {
                  sourceType: 'id',
                  type,
                  id: mediaId,
                  parameters,
                }
              : url
              ? {
                  sourceType: 'url',
                  type,
                  url,
                  parameters,
                }
              : file
              ? {
                  type,
                  sourceType: 'file',
                  parameters: {
                    ...parameters,
                    total_bytes: String(fileSize),
                    media_type: fileType,
                  },
                  fileData: file,
                  fileInfo: {
                    contentType: fileType,
                    knownLength: fileSize,
                  },
                  assetTag,
                }
              : (null as never),
          }),
        ];
      },
    }[name]
  );

/**
 * Send a photo by direct message or attach it to a tweet
 * @category Component
 * @props {@link MediaProps}
 * @guides Check official [guide](https://developer.twitter.com/en/docs/twitter-api/v1/media/upload-media/overview).
 */
export const Photo: TwitterComponent<
  MediaProps,
  UnitSegment<TwitterSegmentValue>
> = makeMediaComponent('Photo', 'photo');

/**
 * Send a video by direct message or attach it to a tweet
 * @category Component
 * @props {@link MediaProps}
 * @guides Check official [guide](https://developer.twitter.com/en/docs/twitter-api/v1/media/upload-media/overview).
 */
export const Video: TwitterComponent<
  MediaProps,
  UnitSegment<TwitterSegmentValue>
> = makeMediaComponent('Video', 'video');

/**
 * Send an animated gif by direct message or attach it to a tweet
 * @category Component
 * @props {@link MediaProps}
 * @guides Check official [guide](https://developer.twitter.com/en/docs/twitter-api/v1/media/upload-media/overview).
 */
export const AnimatedGif: TwitterComponent<
  MediaProps,
  UnitSegment<TwitterSegmentValue>
> = makeMediaComponent('AnimatedGif', 'animated_gif');
