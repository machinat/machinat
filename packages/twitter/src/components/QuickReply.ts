import { makePartSegment, PartSegment } from '@machinat/core/renderer';
import { makeTwitterComponent } from '../utils';
import { TwitterComponent } from '../types';

/**
 * @category Props
 */
export type QuickReplyProps = {
  /** The text label displayed on the button face. Label text is returned as the user's message response. String, max length of 36 characters including spaces. Values with URLs are not allowed and will return an error */
  label: string;
  /** Optional description text displayed under label text. All options must have this property defined if property is present in any option. Text is auto-wrapped and will display on a max of two lines and supports n for controling line breaks. Description text is not include in the user's message response. String, max length of 72 characters including spaces */
  description?: string;
  /** Metadata that will be sent back in the webhook request. String, max length of 1,000 characters including spaces */
  metadata?: string;
};

/**
 * QuickReply a tweet
 * @category Component
 * @props {@link QuickReplyProps}
 * @guides Check official [guide](https://developer.twitter.com/en/docs/twitter-api/v1/direct-messages/quick-replies/overview).
 */
export const QuickReply: TwitterComponent<
  QuickReplyProps,
  PartSegment<unknown>
> = makeTwitterComponent(function QuickReply(node, path) {
  const { label, description, metadata } = node.props;

  return [makePartSegment(node, path, { label, description, metadata })];
});
