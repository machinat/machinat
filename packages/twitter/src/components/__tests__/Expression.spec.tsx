import Sociably from '@sociably/core';
import { isNativeType } from '@sociably/core/utils';
import Renderer from '@sociably/core/renderer';
import TwitterChat from '../../Chat';
import { Photo } from '../Media';
import { Typing } from '../Typing';
import { QuickReply } from '../QuickReply';
import { Expression } from '../Expression';

const renderer = new Renderer('twitter', async (node, path) => [
  { type: 'text', path, node, value: node.props.children },
  { type: 'break', path, node, value: null },
]);
const render = (element) => renderer.render(element, null, null);

it('is a valid Component', () => {
  expect(typeof Expression).toBe('function');
  expect(isNativeType(<Expression>foo</Expression>)).toBe(true);
  expect(Expression.$$platform).toBe('twitter');
});

test('rendering', async () => {
  const segments = await render(
    <Expression customProfileId="11111">
      <p>Hello</p>
      <p>World</p>
    </Expression>
  );
  expect(segments).toMatchSnapshot();
  expect(
    (segments as any).map(({ value: { request, accomplishRequest } }) =>
      accomplishRequest(new TwitterChat('12345', '67890'), request, null)
    )
  ).toMatchInlineSnapshot(`
    Array [
      Object {
        "href": "1.1/direct_messages/events/new.json",
        "method": "POST",
        "parameters": Object {
          "event": Object {
            "message_create": Object {
              "custom_profile_id": "11111",
              "message_data": Object {
                "attachment": undefined,
                "text": "Hello",
              },
              "target": Object {
                "recipient_id": "67890",
              },
            },
            "type": "message_create",
          },
        },
      },
      Object {
        "href": "1.1/direct_messages/events/new.json",
        "method": "POST",
        "parameters": Object {
          "event": Object {
            "message_create": Object {
              "custom_profile_id": "11111",
              "message_data": Object {
                "attachment": undefined,
                "text": "World",
              },
              "target": Object {
                "recipient_id": "67890",
              },
            },
            "type": "message_create",
          },
        },
      },
    ]
  `);
});

test('rendering with quick replies', async () => {
  const segments = await render(
    <Expression
      quickReplies={
        <>
          <QuickReply label="foo" description="FOOOOOOO" metadata="FOO" />
          <QuickReply label="bar" description="BAAAAAAR" metadata="BAR" />
          <QuickReply label="baz" description="BAAAAAAZ" metadata="BAZ" />
        </>
      }
    >
      <p>Hello World</p>
      <p>Choose One</p>
    </Expression>
  );
  expect(segments).toMatchSnapshot();
  expect(
    (segments as any).map(({ value: { request, accomplishRequest } }) =>
      accomplishRequest(new TwitterChat('12345', '67890'), request, null)
    )
  ).toMatchInlineSnapshot(`
    Array [
      Object {
        "href": "1.1/direct_messages/events/new.json",
        "method": "POST",
        "parameters": Object {
          "event": Object {
            "message_create": Object {
              "custom_profile_id": undefined,
              "message_data": Object {
                "attachment": undefined,
                "text": "Hello World",
              },
              "target": Object {
                "recipient_id": "67890",
              },
            },
            "type": "message_create",
          },
        },
      },
      Object {
        "href": "1.1/direct_messages/events/new.json",
        "method": "POST",
        "parameters": Object {
          "event": Object {
            "message_create": Object {
              "custom_profile_id": undefined,
              "message_data": Object {
                "attachment": undefined,
                "quick_reply": Object {
                  "options": Array [
                    Object {
                      "description": "FOOOOOOO",
                      "label": "foo",
                      "metadata": "FOO",
                    },
                    Object {
                      "description": "BAAAAAAR",
                      "label": "bar",
                      "metadata": "BAR",
                    },
                    Object {
                      "description": "BAAAAAAZ",
                      "label": "baz",
                      "metadata": "BAZ",
                    },
                  ],
                  "type": "options",
                },
                "text": "Choose One",
              },
              "target": Object {
                "recipient_id": "67890",
              },
            },
            "type": "message_create",
          },
        },
      },
    ]
  `);
});

test('rendering with media content', async () => {
  const segments = await render(
    <Expression quickReplies={<QuickReply label="foo" />}>
      <p>Hello</p>
      <Photo shared url="http://foo.bar/baz.png" />
    </Expression>
  );
  expect(segments).toMatchSnapshot();

  const chat = new TwitterChat('12345', '67890');
  expect(
    (segments as any).map(({ value: { request, accomplishRequest } }, i) =>
      accomplishRequest(chat, request, i === 1 ? ['11111'] : null)
    )
  ).toMatchInlineSnapshot(`
    Array [
      Object {
        "href": "1.1/direct_messages/events/new.json",
        "method": "POST",
        "parameters": Object {
          "event": Object {
            "message_create": Object {
              "custom_profile_id": undefined,
              "message_data": Object {
                "attachment": undefined,
                "text": "Hello",
              },
              "target": Object {
                "recipient_id": "67890",
              },
            },
            "type": "message_create",
          },
        },
      },
      Object {
        "href": "1.1/direct_messages/events/new.json",
        "method": "POST",
        "parameters": Object {
          "event": Object {
            "message_create": Object {
              "custom_profile_id": undefined,
              "message_data": Object {
                "attachment": Object {
                  "media": Object {
                    "id": "11111",
                  },
                  "type": "media",
                },
                "quick_reply": Object {
                  "options": Array [
                    Object {
                      "description": undefined,
                      "label": "foo",
                      "metadata": undefined,
                    },
                  ],
                  "type": "options",
                },
                "text": undefined,
              },
              "target": Object {
                "recipient_id": "67890",
              },
            },
            "type": "message_create",
          },
        },
      },
    ]
  `);
});

it('throw if no message to attach quick replies', async () => {
  await expect(
    render(
      <Expression quickReplies={<QuickReply label="foo" />}>{null}</Expression>
    )
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"no message content available to attach quick replies"`
  );
  await expect(
    render(
      <Expression quickReplies={<QuickReply label="foo" />}>
        <Typing />
      </Expression>
    )
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"no message content available to attach quick replies"`
  );
});

it('render null if inner content is empty', async () => {
  await expect(render(<Expression>{null}</Expression>)).resolves.toBe(null);
});
