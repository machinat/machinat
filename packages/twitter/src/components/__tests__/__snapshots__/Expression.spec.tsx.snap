// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering 1`] = `
Array [
  Object {
    "node": <p>
      Hello
    </p>,
    "path": "$#Expression.children:0",
    "type": "unit",
    "value": Object {
      "accomplishRequest": [Function],
      "mediaSources": null,
      "request": Object {
        "href": "1.1/direct_messages/events/new.json",
        "method": "POST",
        "parameters": Object {
          "event": Object {
            "message_create": Object {
              "custom_profile_id": "11111",
              "message_data": Object {
                "attachment": undefined,
                "text": "Hello",
              },
              "target": Object {
                "recipient_id": "",
              },
            },
            "type": "message_create",
          },
        },
      },
      "type": "dm",
    },
  },
  Object {
    "node": <p>
      World
    </p>,
    "path": "$#Expression.children:1",
    "type": "unit",
    "value": Object {
      "accomplishRequest": [Function],
      "mediaSources": null,
      "request": Object {
        "href": "1.1/direct_messages/events/new.json",
        "method": "POST",
        "parameters": Object {
          "event": Object {
            "message_create": Object {
              "custom_profile_id": "11111",
              "message_data": Object {
                "attachment": undefined,
                "text": "World",
              },
              "target": Object {
                "recipient_id": "",
              },
            },
            "type": "message_create",
          },
        },
      },
      "type": "dm",
    },
  },
]
`;

exports[`rendering with media content 1`] = `
Array [
  Object {
    "node": <p>
      Hello
    </p>,
    "path": "$#Expression.children:0",
    "type": "unit",
    "value": Object {
      "accomplishRequest": [Function],
      "mediaSources": null,
      "request": Object {
        "href": "1.1/direct_messages/events/new.json",
        "method": "POST",
        "parameters": Object {
          "event": Object {
            "message_create": Object {
              "custom_profile_id": undefined,
              "message_data": Object {
                "attachment": undefined,
                "text": "Hello",
              },
              "target": Object {
                "recipient_id": "",
              },
            },
            "type": "message_create",
          },
        },
      },
      "type": "dm",
    },
  },
  Object {
    "node": <Photo
      shared={true}
      url="http://foo.bar/baz.png"
    />,
    "path": "$#Expression.children:1",
    "type": "unit",
    "value": Object {
      "accomplishRequest": [Function],
      "mediaSources": Array [
        Object {
          "parameters": Object {
            "additional_owners": undefined,
            "media_category": "dm_image",
            "shared": "true",
          },
          "sourceType": "url",
          "type": "photo",
          "url": "http://foo.bar/baz.png",
        },
      ],
      "request": Object {
        "href": "1.1/direct_messages/events/new.json",
        "method": "POST",
        "parameters": Object {
          "event": Object {
            "message_create": Object {
              "custom_profile_id": undefined,
              "message_data": Object {
                "attachment": Object {
                  "media": Object {
                    "id": "",
                  },
                  "type": "media",
                },
                "quick_reply": Object {
                  "options": Array [
                    Object {
                      "description": undefined,
                      "label": "foo",
                      "metadata": undefined,
                    },
                  ],
                  "type": "options",
                },
                "text": undefined,
              },
              "target": Object {
                "recipient_id": "",
              },
            },
            "type": "message_create",
          },
        },
      },
      "type": "dm",
    },
  },
]
`;

exports[`rendering with quick replies 1`] = `
Array [
  Object {
    "node": <p>
      Hello World
    </p>,
    "path": "$#Expression.children:0",
    "type": "unit",
    "value": Object {
      "accomplishRequest": [Function],
      "mediaSources": null,
      "request": Object {
        "href": "1.1/direct_messages/events/new.json",
        "method": "POST",
        "parameters": Object {
          "event": Object {
            "message_create": Object {
              "custom_profile_id": undefined,
              "message_data": Object {
                "attachment": undefined,
                "text": "Hello World",
              },
              "target": Object {
                "recipient_id": "",
              },
            },
            "type": "message_create",
          },
        },
      },
      "type": "dm",
    },
  },
  Object {
    "node": <p>
      Choose One
    </p>,
    "path": "$#Expression.children:1",
    "type": "unit",
    "value": Object {
      "accomplishRequest": [Function],
      "mediaSources": null,
      "request": Object {
        "href": "1.1/direct_messages/events/new.json",
        "method": "POST",
        "parameters": Object {
          "event": Object {
            "message_create": Object {
              "custom_profile_id": undefined,
              "message_data": Object {
                "attachment": undefined,
                "quick_reply": Object {
                  "options": Array [
                    Object {
                      "description": "FOOOOOOO",
                      "label": "foo",
                      "metadata": "FOO",
                    },
                    Object {
                      "description": "BAAAAAAR",
                      "label": "bar",
                      "metadata": "BAR",
                    },
                    Object {
                      "description": "BAAAAAAZ",
                      "label": "baz",
                      "metadata": "BAZ",
                    },
                  ],
                  "type": "options",
                },
                "text": "Choose One",
              },
              "target": Object {
                "recipient_id": "",
              },
            },
            "type": "message_create",
          },
        },
      },
      "type": "dm",
    },
  },
]
`;
